tests/lms/views/decorators/test_h_api.py

```diff
@@ -145,3 +147,162 @@ def util(self, patch):
+        # create another group for the same course.
+        create_course_group(pyramid_request, mock.sentinel.jwt)
+
+        class CourseGroupMatcher:
+            """An object equal to any other object with matching CourseGroup properties."""
+            def __init__(self, pubid, tool_consumer_instance_guid, context_id):
+                self.pubid = pubid
+                self.tool_consumer_instance_guid = tool_consumer_instance_guid
+                self.context_id = context_id
+
+            def __eq__(self, other):
+                return all((
+                    other.pubid == self.pubid,
+                    other.tool_consumer_instance_guid == self.tool_consumer_instance_guid,
+                    other.context_id == self.context_id,
+                ))
+
+        pyramid_request.db.add.assert_called_once_with(CourseGroupMatcher(
+            pubid="TEST_PUBID",
+            tool_consumer_instance_guid="TEST_GUID",
+            context_id="TEST_CONTEXT",
+        ))
+
+    def test_it_calls_and_returns_the_wrapped_view(self, create_course_group, pyramid_request, wrapped):
+        returned = create_course_group(pyramid_request, mock.sentinel.jwt)
+
+        wrapped.assert_called_once_with(pyramid_request, mock.sentinel.jwt)
+        assert returned == wrapped.return_value
+
+    @pytest.fixture
+    def create_course_group(self, wrapped):
+        # Return the actual wrapper function so that tests can call it directly.
+        return create_course_group(wrapped)
+
+    @pytest.fixture
+    def wrapped(self):
+        """Return the wrapped view function."""
+        return mock.MagicMock()
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.params = {
+            # A valid oauth_consumer_key (matches one for which the
+            # provisioning features are enabled).
+            "oauth_consumer_key": "Hypothesise3f14c1f7e8c89f73cefacdd1d80d0ef",
+            "tool_consumer_instance_guid": "TEST_GUID",
+            "context_id": "TEST_CONTEXT",
+            "roles": "Instructor,urn:lti:instrole:ims/lis/Administrator",
+        }
+        pyramid_request.db = mock.MagicMock()
+        return pyramid_request
+
+    @pytest.fixture
+    def models(self, patch):
+        models = patch("lms.views.decorators.h_api.models")
+
+        def side_effect(**kwargs):
+            return mock.create_autospec(CourseGroup, instance=True, **kwargs)
+        models.CourseGroup.side_effect = side_effect
+
+        models.CourseGroup.get.return_value = None
+
+        return models
+
+    @pytest.fixture
+    def util(self, patch):
+        util = patch("lms.views.decorators.h_api.util")
+        util.generate_group_name.return_value = "TEST_GROUP"
+        util.generate_username.return_value = "TEST_USERNAME"
+        return util
+
+    @pytest.fixture
+    def post(self, patch):
+        post = patch("lms.views.decorators.h_api.requests.post")
```

Comment:
Hah! I was like "don't PATCH when you mean POST" but then I read more carefully!